core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -37,8 +37,9 @@ final class KafkaMetadataLog private (
   log: Log,
   scheduler: Scheduler,
   // This object needs to be thread-safe because it is used by the snapshotting thread to notify the
-  // polling thread when snapshots are created.
-  snapshotIds: ConcurrentSkipListSet[OffsetAndEpoch],
+  // polling thread when snapshots are created. Using a Map instead of a Set so that there is no
+  // need to handle NoSuchElementException.
+  snapshotIds: ConcurrentSkipListMap[OffsetAndEpoch, Unit],
```

Comment:
Why does this have to be a map?

Reply:
`ConcurrentSkipListMap` has `lastEntry` and `firstEntry` that return `null` if the map is empty. `ConcurrentSkipListSet` only has `last` and `first` which throws if the set is empty. :dizzy:

Reply:
Throwing is definitely a bad idea. You can't iterator to get the desired behavior?

Reply:
For the sake of having an alternative, what if we replaced the concurrent collection with a TreeMap and a lock? The main contention seems to be around accessing the earliest snapshot id. Perhaps we could use an atomic reference or a volatile for the earliest and latest snapshot ids so that we don't need to go through the lock.

Reply:
We can probably use iterator and reverse-iterator but iterators require `hasNext` follow by `next`. I didn't want to think/research the concurrency and consistency implications.

Reply:
> For the sake of having an alternative, what if we replaced the concurrent collection with a TreeSet and a lock? The main contention seems to be around accessing the earliest snapshot id. Perhaps we could use an atomic reference or a volatile for the earliest and latest snapshot ids so that we don't need to go through the lock.

This would work. Yeah, I think every polling cycle of the `KafkaRaftClient` needs to read the **latest** snapshot in `maybeUpdateOldestSnapshotId`. I am tempted to wait until we/I implement https://issues.apache.org/jira/browse/KAFKA-12155 to see what the access pattern looks like. I am working on this next. What do you think?

Reply:
Sounds fine to me.

Reply:
@jsancio Iterators need to provide consistent behavior or they would not satisfy the Iterator contract. What I mean is: if `hasNext` returns `true`, then `next` has to return an element. It would not be ok for `next` to do something different in that case. So, it seems safe to use here. In any case, not a big deal to use `ConcurrentNavigableMap` since that's what `ConcurrentSkipListSet` uses too. But if we do that, I would suggest creating a simple wrapper that exposes the methods you want.

Reply:
Sounds good @ijuma . I changed the code to use a `ConcurrentSkipListSet`, ascending iterator and descending iterator.